### PR TITLE
feat: molecule serialization

### PR DESCRIPTION
**Pull request by fjchen7**
Note
- I temporarily comment code in some signing class to simplify molecule serialization refactor. These parts will be refactored by in the future PR.
- I didn't remove the old molecule class as they are still being used in signature method. Will also remove them in the fureture PR. 
- The entrance to serialize/deserialize in molecule way is `pack`/`unpack` method in specpfied data structure. However, because not all data structures are custom classes (e.g. byte, byte[], List), we also need a util class `MoleculeConverter` to do serialization job for these types.

### CODE REVIEW

File: ckb-mercury-sdk/src/main/java/org/nervos/mercury/model/req/item/Record.java [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 import org.nervos.ckb.type.Script;
 import org.nervos.ckb.utils.Numeric;
 
 public class Record {
 
-  public static final String SCRIPT_TYPE = "0x01";
+  public static final byte[] SCRIPT_TYPE = new byte[]{0x1};
 
-  public static final String Address_TYPE = "0x00";
+  public static final byte[] Address_TYPE = new byte[]{0x0};
```
Comment (quake) [inline ckb-mercury-sdk/src/main/java/org/nervos/mercury/model/req/item/Record.java:17]:
```suggestion
  public static final byte[] ADDRESS_TYPE = new byte[]{0x0};
```

File: ckb/src/main/java/org/nervos/ckb/transaction/CellCollector.java [hunk 2]
```diff
@@ -157,6 +154,6 @@ private BigInteger calculateTxFee(Transaction transaction, BigInteger feeRate) {
   }
 
   private BigInteger calculateOutputSize(CellOutput cellOutput) {
-    return Utils.ckbToShannon(Serializer.serializeCellOutput(cellOutput).getLength());
+    return Utils.ckbToShannon(cellOutput.pack().toByteArray().length);
```
Comment (quake) [inline ckb/src/main/java/org/nervos/ckb/transaction/CellCollector.java:157]:
```suggestion
    return Utils.ckbToShannon(cellOutput.pack().getSize());
```

File: core/src/main/java/org/nervos/ckb/utils/Calculator.java [hunk 3]
```diff
@@ -10,8 +9,9 @@ public class Calculator {
   private static final int SERIALIZED_TX_OFFSET_BYTE_SIZE = 4;
 
   public static int calculateTransactionSize(Transaction transaction) {
-    Table serializedTx = Serializer.serializeTransaction(transaction);
-    return serializedTx.getLength() + SERIALIZED_TX_OFFSET_BYTE_SIZE;
+    byte[] bytes = transaction.pack().toByteArray();
+    System.out.println(Numeric.toHexString(bytes));
```
Comment (quake) [inline core/src/main/java/org/nervos/ckb/utils/Calculator.java:13]: please remove this line

File: core/src/main/java/org/nervos/ckb/utils/MoleculeConverter.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package org.nervos.ckb.utils;
+
+import java.math.BigInteger;
+import java.util.List;
+import org.nervos.ckb.newtype.concrete.*;
+
+public class MoleculeConverter {
+  public static byte[] littleEndianBigInteger(BigInteger in, int length) {
```
Comment (quake) [inline core/src/main/java/org/nervos/ckb/utils/MoleculeConverter.java:8]:
```suggestion
  public static byte[] toByteArrayLittleEndianUnsigned(BigInteger in, int length) {
```

File: core/src/main/java/org/nervos/ckb/utils/address/AddressGenerator.java [hunk 5]
```diff
@@ -91,7 +90,7 @@ public static String generateBech32mFullAddress(Network network, Script script)
         Numeric.hexStringToByteArray(
             TYPE_FULL_WITH_BECH32M
                 + Numeric.toHexStringNoPrefix(script.codeHash)
-                + Serializer.serializeHashType(script.hashType)
+                + String.format("%02x", script.hashType.pack() & 0xFF)
```
Comment (quake) [inline core/src/main/java/org/nervos/ckb/utils/address/AddressGenerator.java:93]: the return type of `pack()` is a `byte` already, no need to do byte `&` with 0xFF
```suggestion
                + String.format("%02x", script.hashType.pack())
```

Comment (fjchen7) [reply core/src/main/java/org/nervos/ckb/utils/address/AddressGenerator.java:93]: This part is the old way to parse address. I didn't take consideration into them in refactor before I plan to replace with a new class to parse in the future.

File: core/src/main/java/org/nervos/ckb/utils/address/AddressParser.java [hunk 6]
```diff
@@ -79,7 +78,11 @@ public static AddressParseResult parse(String address) throws AddressFormatExcep
 
     if (TYPE_FULL_WITH_BECH32M.equals(type)) {
       byte[] codeHash = Numeric.hexStringToByteArray(payload.substring(2, 66));
-      Script.HashType hashType = Serializer.deserializeHashType(payload.substring(66, 68));
+      byte hashTypeByte =
+          (byte)
+              ((Character.digit(payload.charAt(66), 16) << 4)
+                  + Character.digit(payload.charAt(67), 16));
+      Script.HashType hashType = Script.HashType.unpack(hashTypeByte);
```
Comment (quake) [inline core/src/main/java/org/nervos/ckb/utils/address/AddressParser.java:85]:
```suggestion
      Script.HashType hashType = Script.HashType.unpack(Byte.parseByte (payload.substring(66,68), 16));
```

**Review by quake — APPROVED** [review]

**Pull request merged**